internal/drive_upload.Rmd

```diff
@@ -0,0 +1,129 @@
+---
+title: "Playing with `drive_upload()`"
+author: "Lucy D’Agostino McGowan"
+date: "7/14/2017"
+output: github_document
+---
+
+```{r}
+library("googledrive")
+library("testthat")
+```
+
+Let's upload a file.
+
+```{r}
+drive_auth("drive-token.rds")
+x <- drive_upload(system.file("DESCRIPTION"), "test-upload-lucy")
+```
+
+## overwrite using a `dribble`
+
+Let's try to upload a file into that same location, using the `x` dribble object.
+```{r}
+## this should error
+expect_error(y <- drive_upload(system.file("DESCRIPTION"), x),
+             "File already exists"
+)
+```
+
+Let's add set the parameter `overwrite = TRUE`. Now this should work.
+```{r}
+y <- drive_upload(system.file("DESCRIPTION"), x, overwrite = TRUE)
+```
+
+The `id`s for `x` and `y` should be identical. 
+```{r}
+expect_identical(y$id, x$id)
+```
+
+## overwrite using a `name`
+
+```{r}
+y <- drive_upload(system.file("DESCRIPTION"), "test-upload-lucy", overwrite = TRUE)
+```
+
+```{r}
+expect_identical(y$id, x$id)
+```
+
+If we do not set the parameter `overwrite = TRUE`, it should still create the file. Now we will have 2 files in the same location with the same name.
+
+```{r}
+y <- drive_upload(system.file("DESCRIPTION"), "test-upload-lucy")
+```
+
+Now the `id`s should no longer be identical.
+```{r}
+expect_false(x$id == y$id)
+```
+
+```{r}
+drive_find("test-upload-lucy")
+```
+
+If we try to overwrite *again*, we should receive an error, since we don't know which to overwrite.
+
+```{r}
+expect_error(
+  z <- drive_upload(system.file("DESCRIPTION"),
+                    "test-upload-lucy",
+                    overwrite = TRUE),
+  "Path to overwrite is not unique"
+)
+```
+
+If we use the `x` or `y` `dribble`, however, we should still be able to overwrite.
+
+```{r}
+z <- drive_upload(system.file("DESCRIPTION"), x, overwrite = TRUE)
+```
+
+```{r}
+expect_equal(z$id, x$id)
+```
+
+## folders
+
+Let's make sure this plays nice with folders
+```{r}
+folder <- drive_mkdir("test-folder-lucy")
+```
+
+```{r}
+a <- drive_upload(system.file("DESCRIPTION"), folder, "test-upload-lucy")
+```
+Let's make sure `a` is actually in `folder`.
+```{r}
+a_parent <- unlist(a$files_resource[[1]]$parents)
+expect_equal(a_parent, folder$id)
+```
+
+```{r}
+b <- drive_upload(system.file("DESCRIPTION"), a, overwrite = TRUE)
+```
+
+`b` should also exist in `folder`.
+
+```{r}
+b_parent <- unlist(b$files_resource[[1]]$parents)
+expect_equal(b_parent, folder$id)
+```
+
+If I try to give it a new name, it should error.
+```{r}
+expect_error(
+  c <- drive_upload(system.file("DESCRIPTION"), b, name = "new_name", overwrite = TRUE),
+  "Requested parent folder does not exist"
+)
+```
+*hmm is this a sensible error? it definitely makes sense in some circumstances but...*
```

Comment:
I can't hold all the scenarios in my head right now, but trust you are working through them.

I'll just restate one of my original assertions, i.e. that this should tell you the folder "jenny-fiddle" doesn't exist:

```
drive_upload(R.home('doc/BioC_mirrors.csv'), "jenny-fiddle", name = "something")
```

I worried because "never really erroring if we don't have a folder " suggests somehow we wouldn't catch such a thing. But I'm probably just not following it all.

I continue to think that `drive_upload()` should NOT support "new media AND new name" at the same time.